transformer_lens/HookedTransformer.py

```diff
@@ -1323,6 +1327,12 @@ def set_use_hook_mlp_in(self, use_hook_mlp_in: bool):
         """
         self.cfg.use_hook_mlp_in = use_hook_mlp_in
 
+    def set_use_attn_in(self, use_attn_in: bool):
+        """
+        Toggles whether to allow editing of inputs to each attention head.
+        """
+        self.cfg.use_attn_in = use_attn_in
```

Comment:
if I understand it correctly, the hook inside `model.add_hook("blocks.0.hook_q_input", ...)` will receive a different tensor shape before/after `model.set_use_attn_in(True)`?

if so, it would be nice to check in this method that no affected hooks are active before calling this method (and if so, to show a warning saying ``Please use `model.set_use_attn_in()` before adding any hooks, or after calling `model.reset_hooks()` - this method changes the tensor shape received by your hooks and you might experience unexpected errors when reusing old input hooks.``)

maybe even a few tests would be nice 😅

Reply:
No, I don't think that this is correct. I think that `blocks.0.hook_q_input` will receive a 4D tensor whether `hook_attn_in` is on or off (note that the old comment annotations of a 3D shape were wrong)

What happens is that we either add the fourth dimension before the splitting into Q, K and V (if hook_attn_in is on) or we do the dimension adding as we split into Q, K and V (if hook_attn_in is off)

I will add all of this as a comment, thanks for pointing out that it is confusing!

Reply:
ah, thanks for explanation, number of expected dimensions souds like something to be covered by explicit assertion instead of a potentially incorrect comment (or at least some test cases, potentially in the future if someone is waiting for this PR)

Reply:
Yep! After writing this, I then made better tests based on @JayBaileyCS's comments. 

E.g I think this is done in [this PR](https://github.com/ArthurConmy/TransformerLens/blob/fc232b300c9849be32a9b59b6f4e09e15bcc6577/tests/unit/test_hooks.py#L158#L167)